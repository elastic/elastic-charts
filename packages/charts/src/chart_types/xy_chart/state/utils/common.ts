/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { LegendItem } from '../../../../common/legend';
import { getDistance, Rotation } from '../../../../utils/common';
import { Point } from '../../../../utils/point';
import { BasicSeriesSpec, SeriesType } from '../../utils/specs';
import { GeometriesCounts } from './types';

/** @internal */
export const MAX_ANIMATABLE_BARS = 300;
/** @internal */
export const MAX_ANIMATABLE_LINES_AREA_POINTS = 600;

/** @internal */
export function isHorizontalRotation(chartRotation: Rotation) {
  return chartRotation === 0 || chartRotation === 180;
}

/** @internal */
export function isVerticalRotation(chartRotation: Rotation) {
  return chartRotation === -90 || chartRotation === 90;
}
/**
 * Check if a specs map contains only line or area specs
 * @param specs Map<SpecId, BasicSeriesSpec>
 * @internal
 */
export function isLineAreaOnlyChart(specs: BasicSeriesSpec[]) {
  return !specs.some((spec) => spec.seriesType === SeriesType.Bar);
}

/** @internal */
export function isChartAnimatable(geometriesCounts: GeometriesCounts, animationEnabled: boolean): boolean {
  if (!animationEnabled) {
    return false;
  }
  const { bars, linePoints, areasPoints } = geometriesCounts;
  const isBarsAnimatable = bars <= MAX_ANIMATABLE_BARS;
  const isLinesAndAreasAnimatable = linePoints + areasPoints <= MAX_ANIMATABLE_LINES_AREA_POINTS;
  return isBarsAnimatable && isLinesAndAreasAnimatable;
}

/** @internal */
export function isAllSeriesDeselected(legendItems: LegendItem[]): boolean {
  // eslint-disable-next-line no-restricted-syntax
  for (const legendItem of legendItems) {
    if (!legendItem.isSeriesHidden) {
      return false;
    }
  }
  return true;
}

/**
 * Sorts points in order from closest to farthest from cursor
 * @internal
 */
export const sortClosestToPoint =
  (cursor: Point) =>
  (a: Point, b: Point): number => {
    return getDistance(cursor, a) - getDistance(cursor, b);
  };
