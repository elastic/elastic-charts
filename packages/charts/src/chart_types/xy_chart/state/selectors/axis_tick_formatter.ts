/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { createCustomCachedSelector } from '../../../../state/create_selector';
import { getSettingsSpecSelector } from '../../../../state/selectors/get_settings_spec';
import { Rotation } from '../../../../utils/common';
import { SpecId } from '../../../../utils/ids';
import { defaultTickFormatter, isXDomain } from '../../utils/axis_utils';
import { groupBy } from '../../utils/group_data_series';
import { AxisSpec } from '../../utils/specs';
import { getScaleConfigsFromSpecsSelector } from './get_api_scale_configs';
import { getAxisSpecsSelector, getSeriesSpecsSelector } from './get_specs';

/** @internal */
export type AxisLabelFormatter<V = unknown> = (value: V) => string;

/** @internal */
export type AxisLabelFormatters = { x: Map<SpecId, AxisLabelFormatter>; y: Map<SpecId, AxisLabelFormatter> };

/** @internal */
export const getAxisTickLabelFormatter = createCustomCachedSelector(
  [getSeriesSpecsSelector, getAxisSpecsSelector, getSettingsSpecSelector, getScaleConfigsFromSpecsSelector],
  (seriesSpecs, axesSpecs, { rotation }, scaleConfigs): AxisLabelFormatters => {
    const seriesByGroupId = groupBy(seriesSpecs, ['groupId'], false);
    const axesByGroupId = groupBy(axesSpecs, ['groupId'], false);
    const groupIds = [...new Set([...Object.keys(seriesByGroupId), ...Object.keys(axesByGroupId)])];
    const { timeZone } = scaleConfigs.x;
    // we need to do that by groupId to find the first Y spec formatter fallback
    return groupIds.reduce<AxisLabelFormatters>(
      (acc, groupId) => {
        const ySpecDataFormatter = (seriesByGroupId[groupId] ?? []).find(({ tickFormat }) => tickFormat)?.tickFormat;
        const axes = groupAxesByCartesianCoords(axesByGroupId[groupId] ?? [], rotation);
        axes.x.forEach((spec) => {
          acc.x.set(spec.id, (v) => (spec?.labelFormat ?? spec?.tickFormat ?? defaultTickFormatter)(v, { timeZone }));
        });
        axes.y.forEach((spec) => {
          acc.y.set(spec.id, (v) =>
            (spec.labelFormat ?? spec.tickFormat ?? ySpecDataFormatter ?? defaultTickFormatter)(v, {}),
          );
        });
        return acc;
      },
      { x: new Map(), y: new Map() },
    );
  },
);

function groupAxesByCartesianCoords(sameGroupAxes: AxisSpec[], chartRotation: Rotation = 0) {
  return sameGroupAxes.reduce<{ x: AxisSpec[]; y: AxisSpec[] }>(
    (acc, spec) => {
      acc[isXDomain(spec.position, chartRotation) ? 'x' : 'y'].push(spec);
      return acc;
    },
    { x: [], y: [] },
  );
}
