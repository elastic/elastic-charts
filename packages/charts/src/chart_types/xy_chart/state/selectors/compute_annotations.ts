/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { createCustomCachedSelector } from '../../../../state/create_selector';
import { getChartThemeSelector } from '../../../../state/selectors/get_chart_theme';
import { getSettingsSpecSelector } from '../../../../state/selectors/get_settings_spec';
import { AxisId } from '../../../../utils/ids';
import { computeAnnotationDimensions } from '../../annotations/utils';
import { computeSeriesGeometriesSelector } from './compute_series_geometries';
import { computeSmallMultipleScalesSelector } from './compute_small_multiple_scales';
import { getAxesStylesSelector } from './get_axis_styles';
import { getAxisSpecsSelector, getAnnotationSpecsSelector } from './get_specs';
import { isHistogramModeEnabledSelector } from './is_histogram_mode_enabled';

const getAxisStyleGetter = createCustomCachedSelector(
  [getAxesStylesSelector, getChartThemeSelector],
  (axisStyles, { axes }) =>
    (id: AxisId = '') =>
      axisStyles.get(id) ?? axes,
);

/** @internal */
export const computeAnnotationDimensionsSelector = createCustomCachedSelector(
  [
    getAnnotationSpecsSelector,
    getSettingsSpecSelector,
    computeSeriesGeometriesSelector,
    getAxisSpecsSelector,
    isHistogramModeEnabledSelector,
    computeSmallMultipleScalesSelector,
    getAxisStyleGetter,
  ],
  computeAnnotationDimensions,
);
