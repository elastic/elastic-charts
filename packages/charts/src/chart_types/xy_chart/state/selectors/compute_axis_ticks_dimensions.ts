/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { createCustomCachedSelector } from '../../../../state/create_selector';
import { getChartThemeSelector } from '../../../../state/selectors/get_chart_theme';
import { getSettingsSpecSelector } from '../../../../state/selectors/get_settings_specs';
import { CanvasTextBBoxCalculator } from '../../../../utils/bbox/canvas_text_bbox_calculator';
import { AxisId } from '../../../../utils/ids';
import { axisViewModel, AxisViewModel, hasDuplicateAxis, defaultTickFormatter } from '../../utils/axis_utils';
import { computeSeriesDomainsSelector } from './compute_series_domains';
import { countBarsInClusterSelector } from './count_bars_in_cluster';
import { getAxesStylesSelector } from './get_axis_styles';
import { getBarPaddingsSelector } from './get_bar_paddings';
import { getAxisSpecsSelector, getSeriesSpecsSelector } from './get_specs';
import { isHistogramModeEnabledSelector } from './is_histogram_mode_enabled';

/** @internal */
export const computeAxisTicksDimensionsSelector = createCustomCachedSelector(
  [
    getBarPaddingsSelector,
    isHistogramModeEnabledSelector,
    getAxisSpecsSelector,
    getChartThemeSelector,
    getSettingsSpecSelector,
    computeSeriesDomainsSelector,
    countBarsInClusterSelector,
    getSeriesSpecsSelector,
    getAxesStylesSelector,
  ],
  (
    barsPadding,
    isHistogramMode,
    axesSpecs,
    chartTheme,
    settingsSpec,
    seriesDomainsAndData,
    totalBarsInCluster,
    seriesSpecs,
    axesStyles,
  ): Map<AxisId, AxisViewModel> => {
    const { xDomain, yDomains } = seriesDomainsAndData;
    const fallBackTickFormatter = seriesSpecs.find(({ tickFormat }) => tickFormat)?.tickFormat ?? defaultTickFormatter;
    const bboxCalculator = new CanvasTextBBoxCalculator();
    const axesTicksDimensions: Map<AxisId, AxisViewModel> = new Map();
    axesSpecs.forEach((axisSpec) => {
      const { id } = axisSpec;
      const axisStyle = axesStyles.get(id) ?? chartTheme.axes;
      const dimensions = axisViewModel(
        axisSpec,
        xDomain,
        yDomains,
        totalBarsInCluster,
        bboxCalculator,
        settingsSpec.rotation,
        axisStyle,
        fallBackTickFormatter,
        barsPadding,
        isHistogramMode,
      );
      if (
        dimensions &&
        (!settingsSpec.hideDuplicateAxes || !hasDuplicateAxis(axisSpec, dimensions, axesTicksDimensions, axesSpecs))
      ) {
        axesTicksDimensions.set(id, dimensions);
      }
    });
    bboxCalculator.destroy();
    return axesTicksDimensions;
  },
);
