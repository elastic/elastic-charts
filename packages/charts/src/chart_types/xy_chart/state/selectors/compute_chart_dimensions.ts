/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { createCustomCachedSelector } from '../../../../state/create_selector';
import { getChartContainerDimensionsSelector } from '../../../../state/selectors/get_chart_container_dimensions';
import { getChartThemeSelector } from '../../../../state/selectors/get_chart_theme';
import { getSmallMultiplesSpec } from '../../../../state/selectors/get_small_multiples_spec';
import { computeChartDimensions, ChartDimensions } from '../../utils/dimensions';
import { computeAxisTicksDimensionsSelector } from './compute_axis_ticks_dimensions';
import { getAxesStylesSelector } from './get_axis_styles';
import { getAxisSpecsSelector } from './get_specs';

/** @internal */
export const computeChartDimensionsSelector = createCustomCachedSelector(
  [
    getChartContainerDimensionsSelector,
    getChartThemeSelector,
    computeAxisTicksDimensionsSelector,
    getAxisSpecsSelector,
    getAxesStylesSelector,
    getSmallMultiplesSpec,
  ],
  (chartContainerDimensions, chartTheme, axesTicksDimensions, axesSpecs, axesStyles, smSpec): ChartDimensions =>
    computeChartDimensions(
      chartContainerDimensions,
      chartTheme,
      axesTicksDimensions,
      axesStyles,
      axesSpecs,
      smSpec && smSpec[0],
    ),
);
