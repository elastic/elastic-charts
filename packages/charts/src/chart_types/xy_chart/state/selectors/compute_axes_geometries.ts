/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { createCustomCachedSelector } from '../../../../state/create_selector';
import { getChartThemeSelector } from '../../../../state/selectors/get_chart_theme';
import { getSettingsSpecSelector } from '../../../../state/selectors/get_settings_specs';
import { getAxesGeometries, AxisGeometry, defaultTickFormatter } from '../../utils/axis_utils';
import { computeAxisTicksDimensionsSelector } from './compute_axis_ticks_dimensions';
import { computeChartDimensionsSelector } from './compute_chart_dimensions';
import { computeSeriesDomainsSelector } from './compute_series_domains';
import { computeSmallMultipleScalesSelector } from './compute_small_multiple_scales';
import { countBarsInClusterSelector } from './count_bars_in_cluster';
import { getAxesStylesSelector } from './get_axis_styles';
import { getBarPaddingsSelector } from './get_bar_paddings';
import { getAxisSpecsSelector, getSeriesSpecsSelector } from './get_specs';
import { isHistogramModeEnabledSelector } from './is_histogram_mode_enabled';

/** @internal */
export const computeAxesGeometriesSelector = createCustomCachedSelector(
  [
    computeChartDimensionsSelector,
    getChartThemeSelector,
    getSettingsSpecSelector,
    getAxisSpecsSelector,
    computeAxisTicksDimensionsSelector,
    getAxesStylesSelector,
    computeSeriesDomainsSelector,
    countBarsInClusterSelector,
    isHistogramModeEnabledSelector,
    getBarPaddingsSelector,
    getSeriesSpecsSelector,
    computeSmallMultipleScalesSelector,
  ],
  (
    chartDimensions,
    chartTheme,
    settingsSpec,
    axesSpecs,
    axesTicksDimensions,
    axesStyles,
    seriesDomainsAndData,
    totalBarsInCluster,
    isHistogramMode,
    barsPadding,
    seriesSpecs,
    smScales,
  ): AxisGeometry[] => {
    const fallBackTickFormatter = seriesSpecs.find(({ tickFormat }) => tickFormat)?.tickFormat ?? defaultTickFormatter;
    const { xDomain, yDomains } = seriesDomainsAndData;

    return getAxesGeometries(
      chartDimensions,
      chartTheme,
      settingsSpec.rotation,
      axesSpecs,
      axesTicksDimensions,
      axesStyles,
      xDomain,
      yDomains,
      smScales,
      totalBarsInCluster,
      isHistogramMode,
      fallBackTickFormatter,
      barsPadding,
    );
  },
);
