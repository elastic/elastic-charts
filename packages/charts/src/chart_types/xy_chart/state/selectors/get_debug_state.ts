/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { LegendItem } from '../../../../common/legend';
import { AxisSpec } from '../../../../specs';
import { createCustomCachedSelector } from '../../../../state/create_selector';
import {
  DebugState,
  DebugStateArea,
  DebugStateAxes,
  DebugStateBar,
  DebugStateLegend,
  DebugStateLine,
  DebugStateValue,
} from '../../../../state/types';
import { AreaGeometry, BandedAccessorType, BarGeometry, LineGeometry, PerPanel } from '../../../../utils/geometry';
import { FillStyle, Opacity, StrokeStyle, Visible } from '../../../../utils/themes/theme';
import { isHorizontalAxis } from '../../utils/axis_type_utils';
import { AxisGeometry } from '../../utils/axis_utils';
import { LinesGrid } from '../../utils/grid_lines';
import { computeAxesGeometriesSelector } from './compute_axes_geometries';
import { computeLegendSelector } from './compute_legend';
import { computeSeriesGeometriesSelector } from './compute_series_geometries';
import { computeGridLinesSelector } from './get_grid_lines';
import { getAxisSpecsSelector } from './get_specs';

/**
 * Returns a stringified version of the `debugState`
 * @internal
 */
export const getDebugStateSelector = createCustomCachedSelector(
  [
    computeSeriesGeometriesSelector,
    computeLegendSelector,
    computeAxesGeometriesSelector,
    computeGridLinesSelector,
    getAxisSpecsSelector,
  ],
  ({ geometries }, legend, axes, gridLines, axesSpecs): DebugState => {
    const seriesNameMap = getSeriesNameMap(legend);

    return {
      legend: getLegendState(legend),
      axes: getAxes(axes, axesSpecs, gridLines),
      areas: geometries.areas.map(getAreaState(seriesNameMap)),
      lines: geometries.lines.map(getLineState(seriesNameMap)),
      bars: getBarsState(seriesNameMap, geometries.bars),
    };
  },
);

function getAxes(axesGeoms: AxisGeometry[], axesSpecs: AxisSpec[], gridLines: LinesGrid[]): DebugStateAxes {
  return axesSpecs.reduce<DebugStateAxes>(
    (acc, { position, title, id }) => {
      const geom = axesGeoms.find(({ axis }) => axis.id === id);
      if (!geom) {
        return acc;
      }

      const visibleTicks = geom.visibleTicks.filter(({ label }) => label !== '');
      const labels = visibleTicks.map(({ label }) => label);
      const values = visibleTicks.map(({ value }) => value);

      const gridlines = gridLines
        .flatMap(({ lineGroups }) => lineGroups.find(({ axisId }) => axisId === geom.axis.id)?.lines ?? [])
        .map(({ x1: x, y1: y }) => ({ x, y }));

      acc[isHorizontalAxis(position) ? 'x' : 'y'].push({
        id,
        title,
        position,
        ...(isHorizontalAxis(position)
          ? { labels, values, gridlines }
          : { labels: labels.reverse(), values: values.reverse(), gridlines: gridlines.reverse() }),
      });

      return acc;
    },
    { x: [], y: [] },
  );
}

function getBarsState(
  seriesNameMap: Map<string, string>,
  barGeometries: Array<PerPanel<BarGeometry[]>>,
): DebugStateBar[] {
  const buckets = new Map<string, DebugStateBar>();
  const bars = barGeometries.reduce<BarGeometry[]>((acc, bars) => {
    return [...acc, ...bars.value];
  }, []);
  bars.forEach(
    ({
      color,
      seriesIdentifier: { key },
      seriesStyle: { rect, rectBorder },
      value: { x, y, mark },
      displayValue,
    }: BarGeometry) => {
      const label = displayValue?.text;
      const name = seriesNameMap.get(key) ?? '';
      const bucket: DebugStateBar = buckets.get(key) ?? {
        key,
        name,
        color,
        bars: [],
        labels: [],
        visible: hasVisibleStyle(rect) || hasVisibleStyle(rectBorder),
      };

      bucket.bars.push({ x, y, mark });

      if (label) {
        bucket.labels.push(label);
      }

      buckets.set(key, bucket);

      return buckets;
    },
  );

  return [...buckets.values()];
}

function getLineState(seriesNameMap: Map<string, string>) {
  return ({
    value: {
      line: path,
      points,
      color,
      seriesIdentifier: { key },
      seriesLineStyle,
      seriesPointStyle,
    },
  }: PerPanel<LineGeometry>): DebugStateLine => {
    const name = seriesNameMap.get(key) ?? '';

    return {
      path,
      color,
      key,
      name,
      visible: hasVisibleStyle(seriesLineStyle),
      visiblePoints: hasVisibleStyle(seriesPointStyle),
      points: points.map(({ value: { x, y, mark } }) => ({ x, y, mark })),
    };
  };
}

function getAreaState(seriesNameMap: Map<string, string>) {
  return ({
    value: {
      area: path,
      lines,
      points,
      color,
      seriesIdentifier: { key },
      seriesAreaStyle,
      seriesPointStyle,
      seriesAreaLineStyle,
    },
  }: PerPanel<AreaGeometry>): DebugStateArea => {
    const [y1Path, y0Path] = lines;
    const linePoints = points.reduce<{
      y0: DebugStateValue[];
      y1: DebugStateValue[];
    }>(
      (acc, { value: { accessor, ...value } }) => {
        if (accessor === BandedAccessorType.Y0) {
          acc.y0.push(value);
        } else {
          acc.y1.push(value);
        }

        return acc;
      },
      {
        y0: [],
        y1: [],
      },
    );
    const lineVisible = hasVisibleStyle(seriesAreaLineStyle);
    const visiblePoints = hasVisibleStyle(seriesPointStyle);
    const name = seriesNameMap.get(key) ?? '';

    return {
      path,
      color,
      key,
      name,
      visible: hasVisibleStyle(seriesAreaStyle),
      lines: {
        y0: y0Path
          ? {
              visible: lineVisible,
              path: y0Path,
              points: linePoints.y0,
              visiblePoints,
            }
          : undefined,
        y1: {
          visible: lineVisible,
          path: y1Path,
          points: linePoints.y1,
          visiblePoints,
        },
      },
    };
  };
}

/**
 * returns series key to name mapping
 */
function getSeriesNameMap(legendItems: LegendItem[]): Map<string, string> {
  return legendItems.reduce((acc, { label: name, seriesIdentifiers }) => {
    seriesIdentifiers.forEach(({ key }) => {
      acc.set(key, name);
    });
    return acc;
  }, new Map<string, string>());
}

function getLegendState(legendItems: LegendItem[]): DebugStateLegend {
  const items = legendItems
    .filter(({ isSeriesHidden }) => !isSeriesHidden)
    .map(({ label: name, color, seriesIdentifiers }) => {
      return seriesIdentifiers.map(({ key }) => ({
        key,
        name,
        color,
      }));
    })
    .flat();

  return { items };
}

/**
 * Returns true for styles if they are visible
 * Serves as a catchall for multiple style types
 */
function hasVisibleStyle({
  visible = true,
  fill = '#fff',
  stroke = '#fff',
  strokeWidth = 1,
  opacity = 1,
}: Partial<StrokeStyle & Visible & FillStyle & Opacity>): boolean {
  return Boolean(visible && opacity > 0 && strokeWidth > 0 && fill && stroke);
}
