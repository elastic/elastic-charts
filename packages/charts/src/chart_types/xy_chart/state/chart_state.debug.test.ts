/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { MockSeriesSpec, MockGlobalSpec } from '../../../mocks/specs';
import { MockStore } from '../../../mocks/store/store';
import { ScaleType } from '../../../scales/constants';
import { Rotation } from '../../../utils/common';
import { getDebugStateSelector } from './selectors/get_debug_state';

describe('XYChart - debug state', () => {
  it.each<Rotation>([0, 90, -90, 180])('line chart rotation: %i', (rotation) => {
    const store = MockStore.default({ width: 200, height: 200, top: 0, left: 0 });
    MockStore.addSpecs(
      [
        MockGlobalSpec.settings({ rotation, debugState: true }),
        MockSeriesSpec.line({
          xAccessor: 0,
          yAccessors: [1],
          data: [
            [0, 0],
            [4, 3],
            [10, 10],
          ],
          xScaleType: ScaleType.Linear,
          yScaleType: ScaleType.Linear,
        }),
        MockGlobalSpec.xAxis(
          {
            integersOnly: true,
          },
          rotation,
        ),
        MockGlobalSpec.yAxis(
          {
            integersOnly: true,
          },
          rotation,
        ),
      ],
      store,
    );
    const debugState = getDebugStateSelector(store.getState());
    expect(debugState).toMatchSnapshot();
  });
});
