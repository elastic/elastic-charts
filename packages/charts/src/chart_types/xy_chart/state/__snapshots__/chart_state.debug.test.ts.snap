// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XYChart - debug state line chart rotation: -90 1`] = `
{
  "annotations": [],
  "areas": [],
  "axes": {
    "x": [
      {
        "gridlines": [],
        "id": "xAxis",
        "labels": [
          "0",
          "1",
          "2",
          "3",
          "4",
          "5",
          "6",
          "7",
          "8",
          "9",
          "10",
        ],
        "position": "left",
        "title": undefined,
        "values": [
          0,
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
        ],
      },
    ],
    "y": [
      {
        "gridlines": [],
        "id": "yAxis",
        "labels": [
          "0",
          "1",
          "2",
          "3",
          "4",
          "5",
          "6",
          "7",
          "8",
          "9",
          "10",
        ],
        "position": "bottom",
        "title": undefined,
        "values": [
          0,
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
        ],
      },
    ],
  },
  "bars": [],
  "legend": {
    "items": [
      {
        "color": "#54B399",
        "key": "groupId{__global__}spec{spec1}yAccessor{1}splitAccessors{}",
        "name": "spec1",
      },
    ],
  },
  "lines": [
    {
      "color": "#54B399",
      "key": "groupId{__global__}spec{spec1}yAccessor{1}splitAccessors{}",
      "name": "spec1",
      "path": "M0,159.5L64,112L160,0",
      "points": [
        {
          "mark": null,
          "x": 0,
          "y": 0,
        },
        {
          "mark": null,
          "x": 4,
          "y": 3,
        },
        {
          "mark": null,
          "x": 10,
          "y": 10,
        },
      ],
      "visible": true,
      "visiblePoints": true,
    },
  ],
}
`;

exports[`XYChart - debug state line chart rotation: 0 1`] = `
{
  "annotations": [],
  "areas": [],
  "axes": {
    "x": [
      {
        "gridlines": [],
        "id": "xAxis",
        "labels": [
          "0",
          "1",
          "2",
          "3",
          "4",
          "5",
          "6",
          "7",
          "8",
          "9",
          "10",
        ],
        "position": "bottom",
        "title": undefined,
        "values": [
          0,
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
        ],
      },
    ],
    "y": [
      {
        "gridlines": [],
        "id": "yAxis",
        "labels": [
          "0",
          "1",
          "2",
          "3",
          "4",
          "5",
          "6",
          "7",
          "8",
          "9",
          "10",
        ],
        "position": "left",
        "title": undefined,
        "values": [
          0,
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
        ],
      },
    ],
  },
  "bars": [],
  "legend": {
    "items": [
      {
        "color": "#54B399",
        "key": "groupId{__global__}spec{spec1}yAccessor{1}splitAccessors{}",
        "name": "spec1",
      },
    ],
  },
  "lines": [
    {
      "color": "#54B399",
      "key": "groupId{__global__}spec{spec1}yAccessor{1}splitAccessors{}",
      "name": "spec1",
      "path": "M0,159.5L64,112L160,0",
      "points": [
        {
          "mark": null,
          "x": 0,
          "y": 0,
        },
        {
          "mark": null,
          "x": 4,
          "y": 3,
        },
        {
          "mark": null,
          "x": 10,
          "y": 10,
        },
      ],
      "visible": true,
      "visiblePoints": true,
    },
  ],
}
`;

exports[`XYChart - debug state line chart rotation: 90 1`] = `
{
  "annotations": [],
  "areas": [],
  "axes": {
    "x": [
      {
        "gridlines": [],
        "id": "xAxis",
        "labels": [
          "0",
          "1",
          "2",
          "3",
          "4",
          "5",
          "6",
          "7",
          "8",
          "9",
          "10",
        ],
        "position": "left",
        "title": undefined,
        "values": [
          0,
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
        ],
      },
    ],
    "y": [
      {
        "gridlines": [],
        "id": "yAxis",
        "labels": [
          "0",
          "1",
          "2",
          "3",
          "4",
          "5",
          "6",
          "7",
          "8",
          "9",
          "10",
        ],
        "position": "bottom",
        "title": undefined,
        "values": [
          0,
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
        ],
      },
    ],
  },
  "bars": [],
  "legend": {
    "items": [
      {
        "color": "#54B399",
        "key": "groupId{__global__}spec{spec1}yAccessor{1}splitAccessors{}",
        "name": "spec1",
      },
    ],
  },
  "lines": [
    {
      "color": "#54B399",
      "key": "groupId{__global__}spec{spec1}yAccessor{1}splitAccessors{}",
      "name": "spec1",
      "path": "M0,159.5L64,112L160,0",
      "points": [
        {
          "mark": null,
          "x": 0,
          "y": 0,
        },
        {
          "mark": null,
          "x": 4,
          "y": 3,
        },
        {
          "mark": null,
          "x": 10,
          "y": 10,
        },
      ],
      "visible": true,
      "visiblePoints": true,
    },
  ],
}
`;

exports[`XYChart - debug state line chart rotation: 180 1`] = `
{
  "annotations": [],
  "areas": [],
  "axes": {
    "x": [
      {
        "gridlines": [],
        "id": "xAxis",
        "labels": [
          "0",
          "1",
          "2",
          "3",
          "4",
          "5",
          "6",
          "7",
          "8",
          "9",
          "10",
        ],
        "position": "bottom",
        "title": undefined,
        "values": [
          0,
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
        ],
      },
    ],
    "y": [
      {
        "gridlines": [],
        "id": "yAxis",
        "labels": [
          "0",
          "1",
          "2",
          "3",
          "4",
          "5",
          "6",
          "7",
          "8",
          "9",
          "10",
        ],
        "position": "left",
        "title": undefined,
        "values": [
          0,
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
        ],
      },
    ],
  },
  "bars": [],
  "legend": {
    "items": [
      {
        "color": "#54B399",
        "key": "groupId{__global__}spec{spec1}yAccessor{1}splitAccessors{}",
        "name": "spec1",
      },
    ],
  },
  "lines": [
    {
      "color": "#54B399",
      "key": "groupId{__global__}spec{spec1}yAccessor{1}splitAccessors{}",
      "name": "spec1",
      "path": "M0,159.5L64,112L160,0",
      "points": [
        {
          "mark": null,
          "x": 0,
          "y": 0,
        },
        {
          "mark": null,
          "x": 4,
          "y": 3,
        },
        {
          "mark": null,
          "x": 10,
          "y": 10,
        },
      ],
      "visible": true,
      "visiblePoints": true,
    },
  ],
}
`;
